### ISSUE TITLE
Wrong generics on Tree.transform()

### ISSUE THREAD

**Issue by lukaseder** [javaslang/javaslang#1103]
The current `Tree.transform()` method has a wrong generic type signature:

```java
// Is
<U> U transform(Function<? super Tree<? super T>, ? extends U> f);

// Should be
<U> U transform(Function<? super Tree<T>, ? extends U> f);
```

The following code currently doesn't compile, but would compile if you adapted the signature:

```java
Node<Integer> tree = Tree.of(1, Tree.of(2, Tree.of(3), Tree.of(4)), Tree.of(5, Tree.of(6)));

Function<Tree<Integer>, Tree<String>> toStringInt = i -> i.map(Objects::toString);
Function<Tree<? extends Number>, Tree<String>> toStringNumber = i -> i.map(Objects::toString);
Function<Tree<?>, Tree<String>> toStringAny = i -> i.map(Objects::toString);

System.out.println(tree.transform(toStringInt)); // Doesn't compile
System.out.println(tree.transform(toStringNumber)); // Doesn't compile
System.out.println(tree.transform(toStringAny));
```

I'm not sure what value the `? super T` currently adds... Is there any specific use-case you've had in mind?

I haven't checked other `transform()` methods, but the same thing probably applies.

**Comment by danieldietrich** [comment]
Good morning!
Yes thanks, is saw that yesterday and filed #1100 - the current signature makes no sense.
I think in our case we can even replace `? extends U` with `U`, because method return types are co-variant in Java - but I'm not 100% sure.

**Comment by lukaseder** [comment]
> I think in our case we can even replace ? extends U with U, because method return types are co-variant in Java - but I'm not 100% sure.

Those are differemt beasts. Right now, I can write:

```java
Node<Integer> tree = Tree.of(1, Tree.of(2, Tree.of(3), Tree.of(4)), Tree.of(5, Tree.of(6)));
Function<Tree<?>, Tree<Object>> toStringAny = i -> i.map(Objects::toString);
//                     ^^^^^^ Not String
System.out.println(tree.transform(toStringAny));
```

This wouldn't be possible if you change `? extends U` to `U`

**Comment by danieldietrich** [comment]
Yep, you are right!

**Comment by danieldietrich** [comment]
Duplicate to #1100 but contains additional information

**Comment by danieldietrich** [comment]
Fixed with #1104
